examples/deep_sleep.rs

```diff
@@ -0,0 +1,36 @@
+//! Tests deep sleep
+//!
+//! Prints reset and wakeup reason on start, before enabling mutiple
+//! wakeup sources and start deep sleep. Deep sleep will reset the
+//! device, so after sleep it will restart the application.
+
+use esp_idf_hal::reset::ResetReason;
+use esp_idf_hal::reset::WakeupReason;
+use esp_idf_hal::sleep::DeepSleep;
+use esp_idf_hal::sleep::Sleep;
+use esp_idf_hal::sleep::WakeupSource;
+use std::thread;
+use std::time::Duration;
+
+fn main() -> anyhow::Result<()> {
+    esp_idf_sys::link_patches();
+
+    let reset_reason = ResetReason::get();
+    let wakeup_reason = WakeupReason::get();
+    println!(
+        "start up after reset {:?} wake up reason {:?}",
+        reset_reason, wakeup_reason
+    );
+
+    thread::sleep(Duration::from_secs(2));
+
+    let mut dsleep = DeepSleep::default();
+    dsleep.add_wakeup_source(WakeupSource::Timer {
+        dur: Duration::from_secs(5),
+    })?;
+    println!("{:?}", dsleep);
+
+    dsleep.sleep()?;
+
+    Ok(())
```

Comment:
... will not hit here. See my comment at the end of the review.

Reply:
related to the next